tests/msc2716_test.go

```diff
@@ -0,0 +1,476 @@
+				JSON: []match.JSON{
+					// We're using this weird custom matcher function because we want to iterate the full response
+					// to see all of the events from the response. This way we can use it to easily compare in
+					// the fail error message when the test fails and compare the actual order to the expected order.
+					func(body []byte) error {
+						eventIDsFromResponse, err := getEventIDsFromResponseBody(body)
+						if err != nil {
+							return err
+						}
+
+						// Copy the array by value so we can modify it as we iterate in the foreach loop
+						// We save the full untouched `expectedMessageOrder` for use in the log messages
+						workingExpectedMessageOrder := expectedMessageOrder
+
+						// Match each event from the response in order to the list of expected events
+						matcher := match.JSONArrayEach("chunk", func(r gjson.Result) error {
+							// Find all events in order
+							if len(r.Get("content").Get("body").Str) > 0 {
+								// Pop the next message off the expected list
+								nextEventInOrder := workingExpectedMessageOrder[0]
+								workingExpectedMessageOrder = workingExpectedMessageOrder[1:]
+
+								if r.Get("event_id").Str != nextEventInOrder {
+									return fmt.Errorf("Next event found was %s but expected %s\nActualEvents: %v\nExpectedEvents: %v", r.Get("event_id").Str, nextEventInOrder, eventIDsFromResponse, expectedMessageOrder)
+								}
+							}
+
+							return nil
+						})
+
+						err = matcher(body)
+
+						return err
+					},
+				},
+			})
+		})
+
+		t.Run("Backfilled historical events resolve with proper state", func(t *testing.T) {
+			t.Parallel()
+
+			roomID := as.CreateRoom(t, map[string]interface{}{
+				"preset": "public_chat",
+				"name":   "the hangout spot",
+			})
+			alice.JoinRoom(t, roomID, nil)
+
+			// Create the "live" event we are going to insert our backfilled events next to
+			eventsBefore := createMessagesInRoom(t, alice, roomID, 10)
+			eventBefore := eventsBefore[0]
+			timeAfterEventBefore := time.Now()
+
+			numHistoricalMessages := 6
+			// wait X number of ms to ensure that the timestamp changes enough for each of the messages we try to backfill later
+			time.Sleep(time.Duration(numHistoricalMessages) * time.Millisecond)
+
+			// Create some events after.
+			// Fill up the buffer so we have to scrollback to the inserted history later
+			createMessagesInRoom(t, alice, roomID, 200)
+
+			virtualUserLocalpart := "maria"
+			virtualUserID := fmt.Sprintf("@%s:hs1", virtualUserLocalpart)
+			// Register and join the virtual user
+			ensureRegistered(t, as, virtualUserLocalpart)
+			joinRoom(t, as, virtualUserID, roomID)
+
+			// TODO: Figure out why after joining the virtual user and it gets a 200 OK,
+			// we still see `SynapseError: 403 - User @maria:hs1 not in room !ZIIflEwiKrDFeMEMKV:hs1`
+			// when sending a message. Debugging with an Element interface on top even, shows Maria as joined!
```

Comment:
For reference, here is the Complement summary of the requests flying around. The `/reigster` give a `200` and the `/join` gives `200` but sending a message is still giving a `403`. With the more details below, the error says that the user isn't in the room. But I'm not sure how that is possible if the request shows a `200` and shows the user joined when looking at it in Element (see screenshot below)

```
=== CONT  TestBackfillingHistory
    client.go:310: GET /_matrix/client/r0/sync => 200 OK (32.446706ms)
    client.go:310: POST /_matrix/client/r0/register => 200 OK (19.383793ms)
    client.go:310: POST /_matrix/client/r0/join/!ZIIflEwiKrDFeMEMKV:hs1 => 200 OK (30.548256ms)
    client.go:310: PUT /_matrix/client/r0/rooms/!ZIIflEwiKrDFeMEMKV:hs1/send/m.room.message/msc2716-txn1 => 403 Forbidden (32.646351ms)
=== CONT  TestBackfillingHistory/parallel/Backfilled_historical_events_resolve_with_proper_state
    msc2716_test.go:368: CSAPI.MustDoRaw PUT http://localhost:55593/_matrix/client/r0/rooms/%21ZIIflEwiKrDFeMEMKV:hs1/send/m.room.message/msc2716-txn1?access_token=e5207da5489e95d84f98013afcd89f8256004029cb299a1b4d33a171c622bfdd&prev_event=%24wIL3vhgJOh42ZU8n-upY1KPrTTKn9N3sTJ3jTudGwTU&ts=1615885523884&user_id=%40maria%3Ahs1 returned HTTP 403
```

Here are more details behind the request,

 - The `/register` (200) shows it's doing the proper user `{ "username": "maria" }`
 - The `/join` (200) is working on `&user_id=%40maria%3Ahs1`
 - And the send message request throws a `403` `User @maria:hs1 not in room !ZIIflEwiKrDFeMEMKV:hs1`

```
2021-03-16 09:05:35,137 - synapse.http.servlet - 299 - INFO - POST-423 - Body b'{ "username": "maria" }'
2021-03-16 09:05:35,149 - synapse.handlers.auth - 797 - INFO - POST-423 - Logging in user @maria:hs1 on device ZUAQZFGXCT
2021-03-16 09:05:35,153 - synapse.access.http.8008 - 295 - INFO - POST-423 - ::ffff:172.20.0.1 - 8008 - {@the-bridge-user:hs1} Processed request: 0.017sec/-0.000sec (0.006sec, 0.001sec) (0.002sec/0.009sec/6) 331B 200 "POST /_matrix/client/r0/register?access_token=<redacted> HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-03-16 09:05:35,158 - synapse.http.servlet - 299 - INFO - POST-424 - Body b''
2021-03-16 09:05:35,158 - synapse.http.servlet - 309 - WARNING - POST-424 - Unable to parse JSON: Expecting value: line 1 column 1 (char 0)
2021-03-16 09:05:35,164 - synapse.events.builder - 145 - INFO - POST-424 - event_id_with_max_depth $Lp1Bn5T6xgL3nqKDVvZLQNXbMyGoiF_KNxjelPJN1-U 217
2021-03-16 09:05:35,164 - synapse.events.builder - 190 - INFO - POST-424 - depth 218
2021-03-16 09:05:35,174 - synapse.handlers.message - 1252 - INFO - POST-424 - backfilled False
2021-03-16 09:05:35,184 - synapse.access.http.8008 - 295 - INFO - POST-424 - ::ffff:172.20.0.1 - 8008 - {@maria:hs1} Processed request: 0.027sec/-0.000sec (0.006sec, 0.000sec) (0.004sec/0.004sec/11) 37B 200 "POST /_matrix/client/r0/join/%21ZIIflEwiKrDFeMEMKV:hs1?access_token=<redacted>&user_id=%40maria%3Ahs1 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021-03-16 09:05:35,189 - synapse.http.servlet - 299 - INFO - PUT-425 - Body b'{"body":"Historical 2","m.historical":true,"msgtype":"m.text"}'
2021-03-16 09:05:35,196 - synapse.events.builder - 145 - INFO - PUT-425 - event_id_with_max_depth $wIL3vhgJOh42ZU8n-upY1KPrTTKn9N3sTJ3jTudGwTU 8
2021-03-16 09:05:35,205 - synapse.events.builder - 169 - INFO - PUT-425 - sucessor_event_ids ['$74h1OtvFStbQdkhy1_JLPgHV8HarjAnQP2b-9aWSwTw']
2021-03-16 09:05:35,214 - synapse.events.builder - 183 - INFO - PUT-425 - oldest_successor_depth 9
2021-03-16 09:05:35,214 - synapse.events.builder - 190 - INFO - PUT-425 - depth 9
2021-03-16 09:05:35,215 - synapse.handlers.message - 959 - WARNING - PUT-425 - Denying new event <FrozenEventV3 event_id='$YqV9UccLxs46HuEkgset2ODJN9OSvBCDHSUi7hGRUb4', type='m.room.message', state_key=None> because 403: User @maria:hs1 not in room !ZIIflEwiKrDFeMEMKV:hs1 (None)
2021-03-16 09:05:35,216 - synapse.http.server - 85 - INFO - PUT-425 - <SynapseRequest at 0x7f27d09e7a00 method='PUT' uri='/_matrix/client/r0/rooms/%21ZIIflEwiKrDFeMEMKV:hs1/send/m.room.message/msc2716-txn1?access_token=<redacted>&prev_event=%24wIL3vhgJOh42ZU8n-upY1KPrTTKn9N3sTJ3jTudGwTU&ts=1615885523884&user_id=%40maria%3Ahs1' clientproto='HTTP/1.1' site='8008'> SynapseError: 403 - User @maria:hs1 not in room !ZIIflEwiKrDFeMEMKV:hs1 (None)
2021-03-16 09:05:35,216 - synapse.access.http.8008 - 295 - INFO - PUT-425 - ::ffff:172.20.0.1 - 8008 - {@maria:hs1} Processed request: 0.028sec/-0.000sec (0.003sec, 0.001sec) (0.021sec/0.001sec/3) 94B 403 "PUT /_matrix/client/r0/rooms/%21ZIIflEwiKrDFeMEMKV:hs1/send/m.room.message/msc2716-txn1?access_token=<redacted>&prev_event=%24wIL3vhgJOh42ZU8n-upY1KPrTTKn9N3sTJ3jTudGwTU&ts=1615885523884&user_id=%40maria%3Ahs1 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
```

---

Also hooking up the room to a Matrix client shows that Maria did in fact join.

Reply:
Maybe it's because when we try to backfill the historical message, the `m.room.member` event isn't back in time where the state is trying to be resolved.

Reply:
Seems like the bulk send API is necessary if we're using the `backfilled = true` Synapse implementation which requires reverse-chronological history insertion.

With the single send, we can't insert a room member before the actual message. If we insert a `m.room.member` back in time, a subsequent message has to come before it in time so it wont' resolve that state that came after.

Reply:
Related to https://github.com/matrix-org/synapse/pull/9247#discussion_r595425760

Reply:
bulk endpoint added and historical state allows the historical messages to auth properly!